src/wasm/wasm-type.cpp

```diff
@@ -1347,13 +1635,304 @@ Type TypeBuilder::getTempRttType(size_t i, uint32_t depth) {
+
+  ShapeCanonicalizer(const std::vector<HeapType>& input);
+
+private:
+  using TypeSet = std::unordered_set<HeapType>;
+
+  // The HeapTypes in the type definition graph to canonicalize.
+  const std::vector<HeapType>& input;
+
+  // The partitioning of the input HeapTypes used by Hopcroft's algorithm.
+  std::vector<TypeSet> partitions;
+
+  // Hopcroft's algorithm needs to be able to find the predecessors of a
+  // particular state via a given symbol in the alphabet. We use simple child
+  // indices as the alphabet.
+  size_t alphabetSize = 0;
+  std::unordered_map<HeapType, std::unordered_map<size_t, TypeSet>> preds;
+
+  void initializePredecessors();
+  void initializePartitions();
+  void translatePartitionsToTypes();
+
+  std::vector<HeapType*> getChildren(HeapType type);
+  const TypeSet& getPredsOf(HeapType type, size_t symbol);
+  TypeSet getIntersection(const TypeSet& a, const TypeSet& b);
+  TypeSet getDifference(const TypeSet& a, const TypeSet& b);
+};
+
+ShapeCanonicalizer::ShapeCanonicalizer(const std::vector<HeapType>& input)
+  : input(input) {
+  initializePredecessors();
+  initializePartitions();
+
+  // The Hopcroft's algorithm's list of partitions that may still be
+  // distinguishing partitions. Starts out containing all partitions.
+  std::unordered_set<size_t> distinguishers;
+  for (size_t i = 0; i < partitions.size(); ++i) {
+    distinguishers.insert(i);
+  }
+
+  // Hopcroft's algorithm
+  while (distinguishers.size()) {
+    // Choose a partition that might be able to distinguish between the members
+    // of some other partition.
+    auto distinguishingIndexIt = distinguishers.begin();
+    TypeSet distinguishing = partitions[*distinguishingIndexIt];
+    distinguishers.erase(distinguishingIndexIt);
+    // For each possibly distinguishing transition symbol...
+    for (size_t symbol = 0; symbol < alphabetSize; ++symbol) {
+      // Find all types that reach one of the current distinguishing types via
+      // `symbol`.
+      TypeSet currPreds;
+      for (auto type : distinguishing) {
+        const TypeSet& specificPreds = getPredsOf(type, symbol);
+        currPreds.insert(specificPreds.begin(), specificPreds.end());
+      }
+      // Find partitions that contain some elements that are predecessors of the
+      // current distinguishing partition and some elements that are not
+      // predecessors of the current distinguishing partition.
+      for (size_t distinguishedIndex = 0, end = partitions.size();
+           distinguishedIndex < end;
+           ++distinguishedIndex) {
+        TypeSet& distinguished = partitions[distinguishedIndex];
+        TypeSet intersection = getIntersection(distinguished, currPreds);
+        if (intersection.empty()) {
+          continue;
+        }
+        TypeSet difference = getDifference(distinguished, currPreds);
+        if (difference.empty()) {
+          continue;
+        }
+        // We can split the partition! Replace it with the intersection and add
```

Comment:
this `!` sounds too happy to me... we just found more work we have to reluctantly do! :wink:

Reply:
Hahaha should I make it `// We must bid this partition farewell 😢`